lib/runners/processor/pmd_cpd.rb

```diff
@@ -0,0 +1,191 @@
+
+      return {
+        id: id,
+        path: path,
+        location: location
+      }
+    end
+
+    def create_issue_object(elem_dupli, files)
+      lines = elem_dupli[:lines].to_i
+      tokens = elem_dupli[:tokens].to_i
+      codefragment = elem_dupli.elements['codefragment'].cdatas[0].value
+      fileobjs = files.map { |f| {
+        id: f[:id],
+        path: f[:path].to_s,
+        start_line: f[:location].start_line,
+        start_column: f[:location].start_column,
+        end_line: f[:location].end_line,
+        end_column: f[:location].end_column,
+      }}
+
+      return {
+        lines: lines,
+        tokens: tokens,
+        files: fileobjs,
+        codefragment: codefragment
+      }
+    end
+
+    def option_files
+      Array(config_linter[:files] || DEFAULT_FILES).map { |v| ["--files", v] }.flatten
+    end
+
+    def option_encoding
+      config_linter[:encoding].then { |v| v ? ["--encoding", v] : [] }
+    end
+
+    def option_minimum_tokens
+      (config_linter[:'minimum-tokens'] || DEFAULT_MINIMUM_TOKENS).then { |v| ["--minimum-tokens", v.to_s] }
+    end
+
+    def option_language
+      config_linter[:language].then { |v| v ? ["--language", v] : [] }
+    end
+
+    def option_skip_duplicate_files
+      config_linter[:'skip-duplicate-files'].then { |v| v ? ["--skip-duplicate-files"] : [] }
+    end
+
+    def option_non_recursive
+      config_linter[:'non-recursive'].then { |v| v ? ["--non-recursive"] : [] }
+    end
+
+    def option_skip_lexical_errors
+      config_linter[:'skip-lexical-errors'].then { |v| v ? ["--skip-lexical-errors"] : [] }
+    end
+
+    def option_ignore_annotations
+      config_linter[:'ignore-annotations'].then { |v| v ? ["--ignore-annotations"] : [] }
+    end
+
+    def option_ignore_identifiers
+      config_linter[:'ignore-identifiers'].then { |v| v ? ["--ignore-identifiers"] : [] }
+    end
+
+    def option_ignore_literals
+      config_linter[:'ignore-literals'].then { |v| v ? ["--ignore-literals"] : [] }
+    end
+
+    def option_ignore_usings
+      config_linter[:'ignore-usings'].then { |v| v ? ["--ignore-usings"] : [] }
+    end
+
+    def option_no_skip_blocks
+      config_linter[:'no-skip-blocks'].then { |v| v ? ["--no-skip-blocks"] : [] }
+    end
+
+    def option_skip_blocks_pattern
+      config_linter[:'skip-blocks-pattern'].then { |v| v ? ["--skip-blocks-pattern", v] : [] }
+    end
+
+    def cli_options
+      [
+        *option_encoding,
+        *option_minimum_tokens,
+        *option_language,
+        *option_skip_duplicate_files,
+        *option_non_recursive,
+        *option_skip_lexical_errors,
+        *option_ignore_annotations,
+        *option_ignore_identifiers,
+        *option_ignore_literals,
+        *option_ignore_usings,
+        *option_no_skip_blocks,
+        *option_skip_blocks_pattern,
+        "--format", "xml",
```

Comment:
[suggestion] What about setting `--failOnViolation false`?

```suggestion
        "--format", "xml",
        "--failOnViolation", "false",
```

If doing so, `#run_analyzer` method will be simpler (because `capture3!` can be used):

```ruby
def run_analyzer
  stdout, = capture3!(analyzer_bin, *cli_options)
  Results::Success.new(guid: guid, analyzer: analyzer).tap do |result|
    construct_result(result, stdout)
  end
end
```

`capture3!` method can raise `Shell::ExecError` if the process exits with the non-zero code.

Also, I think the following code is not good. Because `Results::Failure` is returned when `exitstatus` is not `0` or `4`.
`Results::Failure` is due to user's misconfiguration etc. and is not reported to Bugsnag, so we will be hard to notice what happens if unexpected errors occur.

```
if status.success? || status.exitstatus == 4
  Results::Success.new(guid: guid, analyzer: analyzer).tap do |result|
    construct_result(result, stdout, stderr)
  end
else
  Results::Failure.new(guid: guid, analyzer: analyzer, message: "Unexpected error occurred. Please see the analysis log.")
end
```

Reply:
That makes sense. Thank you for the detailed explanation.
Fixed. commit b090cb1fe9f00c829be4f1d976b9665ae0d54b7e

Reply:
Nice! 👏